walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,179 @@
+
+
+            for(String attendee: eV.getAttendees()){
+                if(request.getAttendees().contains(attendee)){
+
+
+                        
+
+                        if(currentAttendees.contains(attendee)){
+                            conflict = true;
+                            continue;
+                            
+                        }
+
+                            
+                        
+
+                        if(previous_endTime < eV.getWhen().start()){
+                            
+                                TimeRange tempTr = new TimeRange(0, 24 * 60);
+                                tempTr = tempTr.fromStartEnd(previous_endTime,eV.getWhen().start() - 1,true);
+                                availableTimes.add(tempTr);
+                        }
+
+
+
+                        if(previous_endTime <  eV.getWhen().end()){
+                            previous_endTime = eV.getWhen().end();
+                        }
+                        
+                        
+
+                        if(overlaps(availableTimes,getTimeBefore(eV))){
+                       
+                        availableTimes.add(getTimeBefore(eV));
+                        }
+
+
+
+
+                        tempConflict = eV.getWhen().end();
+                        
+                        currentAttendees.add(attendee);
+
+                }
+                
+            }
+        }
+
+      
+        
+        availableTimes.add(getTimeAfter(previous_endTime));
+    
+ 
+    return availableTimes;
+    // return results;
+    }
+
+
+    public static ArrayList<Event> loadEvents(Collection<Event> events){
+         ArrayList<Event> result = new ArrayList<Event>();
+
+        for(Event ev : events){
+            result.add(ev);
+        }
+        return result;
+    }
+
+
+   public static TimeRange getTimeAfter(int start){
+       
+        TimeRange tempT = new TimeRange(start, 24 * 60);
+       tempT = tempT.fromStartEnd(start,(24 * 60) - 1,true);
+       return tempT;
+   }
+
+    public static TimeRange getTimeBefore(Event ev){
+       
+        TimeRange tempT = new TimeRange(0, ev.getWhen().start());
+        
+       return tempT;
+   }
+
+   
+
+
+
+    public boolean overlaps(ArrayList<TimeRange> possibleTimes, TimeRange ev){
+        for(TimeRange t : possibleTimes){
+            if(ev.start() == t.start()){
+                return false;
+            }
+
+        }
+        return true;
+    }
+
+
+
+public ArrayList<TimeRange> getGaps(ArrayList<Event> eventCopy){
+    // find possibles gaps
+    ArrayList<TimeRange> result = new ArrayList<TimeRange>();
+
+    for(int i = 1; i < eventCopy.size(); i++){
+    
+      TimeRange tempT = new TimeRange(0, 24 * 60);
+        tempT = tempT.fromStartEnd(eventCopy.get(i - 1).getWhen().end(),eventCopy.get(i).getWhen().start() - 1,true);
+    
+    
+  
+    result.add(tempT);
+    }
+
+    return result;
+  
+
 }
+
+    
+
+  }   
+    
+    
+    
+    
+    
+    
```

Comment:
Hopefully I covered all of the major changes that are needed!  Beyond that, just be sure to clean things up and remove any code that's no longer used.  When all is said and done, you should be left with something like the following:

```java
public final class FindMeetingQuery {
  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
    ArrayList<TimeRange> availableTimes = new ArrayList();
    ArrayList<Event> eventsCopy = copyEventsToArrayList(events);
    eventsCopy.sort(Comparator.comparing(Event::getWhen, TimeRange.ORDER_BY_START));

    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
      return availableTimes; // return empty array
    }

    if (request.getAttendees().size() == 0) {
      availableTimes.add(TimeRange.WHOLE_DAY);
      return availableTimes;
    }

    int previous_endTime = TimeRange.START_OF_DAY;

    for (Event eV : eventsCopy) {
      if (!Collections.disjoint(eV.getAttendees(), request.getAttendees())) {
        if (previous_endTime < eV.getWhen().start()) {
          TimeRange tempTr = TimeRange.fromStartEnd(previous_endTime, eV.getWhen().start(), false);
          availableTimes.add(tempTr);
        }

        if (previous_endTime < eV.getWhen().end()) {
          previous_endTime = eV.getWhen().end();
        }
      }
    }

    if (TimeRange.END_OF_DAY - previous_endTime >= request.getDuration()) {
      availableTimes.add(getTimeAfter(previous_endTime));
    }

    return availableTimes;
  }

  public static ArrayList<Event> copyEventsToArrayList(Collection<Event> events) {
    ArrayList<Event> result = new ArrayList<Event>();
    for (Event ev : events) {
      result.add(ev);
    }
    return result;
  }

  public static TimeRange getTimeAfter(int start) {
    return TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true);
  }
}
```